python/lsst/daf/butler/registry/bridge/monolithic.py

```diff
@@ -168,23 +174,36 @@ def insert(self, refs: Iterable[DatasetIdRef]) -> None:
 
     def forget(self, refs: Iterable[DatasetIdRef]) -> None:
         # Docstring inherited from DatastoreRegistryBridge
-        rows = self._refsToRows(self.check(refs))
-        self._db.delete(self._tables.dataset_location, ["datastore_name", "dataset_id"], *rows)
+        with self._db.transaction():
+            for refs_chunk in chunk_iterable(refs, 50_000):
+                dataset_ids = [ref.id for ref in refs_chunk]
+                where = sqlalchemy.sql.and_(
+                    self._tables.dataset_location.columns.datastore_name == self.datastoreName,
+                    self._tables.dataset_location.columns.dataset_id.in_(dataset_ids),
+                )
+                self._db.deleteWhere(self._tables.dataset_location, where)
 
     def moveToTrash(self, refs: Iterable[DatasetIdRef], transaction: DatastoreTransaction | None) -> None:
         # Docstring inherited from DatastoreRegistryBridge
-        # TODO: avoid self.check() call via queries like
-        #     INSERT INTO dataset_location_trash
-        #         SELECT datastore_name, dataset_id FROM dataset_location
-        #         WHERE datastore_name=? AND dataset_id IN (?);
-        #     DELETE FROM dataset_location
-        #         WHERE datastore_name=? AND dataset_id IN (?);
-        # ...but the Database interface doesn't support those kinds of queries
-        # right now.
-        rows = self._refsToRows(self.check(refs))
+        location = self._tables.dataset_location
+        location_trash = self._tables.dataset_location_trash
         with self._db.transaction():
-            self._db.delete(self._tables.dataset_location, ["datastore_name", "dataset_id"], *rows)
-            self._db.insert(self._tables.dataset_location_trash, *rows)
+            for refs_chunk in chunk_iterable(refs, 50_000):
```

Comment:
Can you add a small comment explaining why deleting N rows and inserting N rows is less efficient than reusing a where clause for the insert and delete?